src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -7,6 +7,18 @@
 
 namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
 {
+    // The implementation of project snapshot manager abstracts over the Roslyn Project (WorkspaceProject)
+    // and information from the host's underlying project system (HostProject), to provide a unified and
+    // immutable view of the underlying project systems.
+    //
+    // The HostProject support all of the configuration that the Razor SDK exposes via the project system
+    // (language version, extensions, named configuration).
+    //
+    // The WorkspaceProject is needed to support our use of Roslyn Compilations for Tag Helpers and other
+    // C# based constructs.
+    //
+    // The somewhat complex task performed here is to unify these two concepts (both independently stateful)
+    // and also to provide a facility for computed state based on these two concepts.
```

Comment:
Think of this PR as a rewrite of this class. The state that this class has to manage is much more sophisticated now.

Reply:
Definitely super sophisiticated I agree. Was a skeptical at first but the more I think about/see what you did, I'm liking it a lot.